core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -117,4 +122,45 @@ private static void deleteFiles(FileIO io, Set<ManifestFile> allManifests) {
           }
         });
   }
+
+  /**
+   * Load a custom catalog implementation.
+   * The catalog must have a no-arg constructor.
+   * If the catalog implements {@link org.apache.hadoop.conf.Configurable},
+   * {@code Configurable.setConf(org.apache.hadoop.conf.Configuration conf)} is called to set Hadoop configuration.
+   * {@code Catalog.initialize(String name, Map<String, String> options)} is called to complete the initialization.
+   * @param catalogName catalog name
+   * @param impl catalog implementation full class name
+   * @param engineOptions configuration options from a compute engine like Spark or Flink to initialize the catalog
+   * @param hadoopConf hadoop configuration if needed
+   * @return initialized catalog object
+   * @throws IllegalArgumentException if no-arg constructor not found or error during initialization
+   */
+  public static Catalog loadCustomCatalog(
+      String catalogName,
+      String impl,
+      Map<String, String> engineOptions,
+      Configuration hadoopConf) {
+    Preconditions.checkNotNull(impl, "Cannot initialize custom Catalog because impl property is not set");
+    DynConstructors.Ctor<Catalog> ctor;
+    try {
+      ctor = DynConstructors.builder(Catalog.class)
+          .impl(impl)
+          .buildChecked();
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException(String.format(
+          "Cannot initialize Catalog, please make sure %s has a no-arg constructor", impl), e);
+    }
+    try {
+      Catalog catalog = ctor.newInstance();
+      if (catalog instanceof Configurable) {
+        ((Configurable) catalog).setConf(hadoopConf);
+      }
+      catalog.initialize(catalogName, engineOptions);
```

Comment:
Can we move configuration out of the try/catch block? It doesn't need to be there.